File: leetcode/131.palindrome-partitioning.py
class Solution(object):
    def __init__(self):
        self.is_palindrome = None

    def pre_build(self, s):

        size = len(s)
        self.is_palindrome = [[False for _ in range(size)] for _ in range(size)]

        for t in range(size):
            i = j = t
            while i >= 0 and j < size and s[i] == s[j]:
                self.is_palindrome[i][j] = True
                i -= 1
                j += 1

            i = t
            j = t + 1
            while i >= 0 and j < size and s[i] == s[j]:
                self.is_palindrome[i][j] = True
                i -= 1
                j += 1

    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """

        self.pre_build(s)
        results = []
        ans = []
        self.helper(s, 0, ans, results)
        return results

    def helper(self, s, start, ans, results):
        
        if start == len(s):
            results.append(list(ans))

        for i in range(start, len(s)):
    
            if not self.is_palindrome[start][i]:
                continue

            ans.append(s[start: i + 1])
            self.helper(s, i + 1, ans, results)
            ans.pop()

File: leetcode/131.palindrome-partitioning.cpp
class Solution {
private:
    vector<vector<bool>> isPalindrome;
    
    void helper(string& s, int start, vector<string>& ans, vector<vector<string>>& res)
    {
        if (start == s.size())
        {
            res.push_back(ans);
            return;
        }
        
        for (auto i = start; i < s.size(); ++i)
        {
            if (isPalindrome[start][i])
            {
                ans.push_back(s.substr(start, i - start + 1));
                helper(s, i + 1, ans, res);
                ans.pop_back();
            }
        }
    }
public:
    vector<vector<string>> partition(string s) {
        int size = s.size();
        isPalindrome.resize(size, vector<bool>(size, false));
        
        int i, j =0;
        for (auto t = 0; t < size; ++t)
        {
            i = j = t;
            while(i >= 0 && j < size && s[i] == s[j])
            {
                isPalindrome[i][j] = true;
                i--;
                j++;
            }
            
            i = t;
            j = t + 1;
            while(i >= 0 && j < size && s[i] == s[j])
            {
                isPalindrome[i][j] = true;
                i--;
                j++;
            }
        }
        
        vector<string> ans;
        vector<vector<string>> res;
        helper(s, 0, ans, res);
        
        return res;
    }
};